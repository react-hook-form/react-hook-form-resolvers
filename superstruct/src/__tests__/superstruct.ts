import {
  object,
  number,
  string,
  optional,
  pattern,
  size,
  union,
  min,
  max,
  Infer,
  define,
  array,
  boolean,
} from 'superstruct';
import { superstructResolver } from '..';

const Password = define('Password', (value, ctx) =>
  value === ctx.branch[0].password);

const schema = object({
  username: size(pattern(string(), /^\w+$/), 3, 30),
  password: pattern(string(), /^[a-zA-Z0-9]{3,30}/),
  repeatPassword: Password,
  accessToken: optional(union([string(), number()])),
  birthYear: optional(max(min(number(), 1900), 2013)),
  email: optional(pattern(string(), /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/)),
  tags: array(string()),
  enabled: boolean(),
});

describe('superstructResolver', () => {
  it('should return values from superstructResolver when validation pass', async () => {
    const data: Infer<typeof schema> = {
      username: 'Doe',
      password: 'Password123',
      repeatPassword: 'Password123',
      birthYear: 2000,
      email: 'john@doe.com',
      tags: ['tag1', 'tag2'],
      enabled: true,
    };

    const result = await superstructResolver(schema)(data, undefined, {
      fields: {},
    });

    expect(result).toEqual({ errors: {}, values: data });
  });

  it('should return a single error from superstructResolver when validation fails', async () => {
    const data = {
      password: '___',
      email: '',
      birthYear: 'birthYear',
    };

    const result = await superstructResolver(schema)(data, undefined, {
      fields: {},
    });

    expect(result).toMatchSnapshot();
  });

  it('should return all the errors from superstructResolver when validation fails with `validateAllFieldCriteria` set to true', async () => {
    const data = {
      password: '___',
      email: '',
      birthYear: 'birthYear',
    };

    const result = await superstructResolver(schema)(data, undefined, {
      fields: {},
      criteriaMode: 'all',
    });

    expect(result).toMatchSnapshot();
  });
});
