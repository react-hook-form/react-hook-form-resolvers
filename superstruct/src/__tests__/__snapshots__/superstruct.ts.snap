// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`superstructResolver should return a single error from superstructResolver when validation fails 1`] = `
Object {
  "errors": Object {
    "birthYear": Object {
      "message": "Expected a number, but received: \\"birthYear\\"",
      "ref": undefined,
      "type": "number",
    },
    "email": Object {
      "message": "Expected a string matching \`/^[\\\\w-\\\\.]+@([\\\\w-]+\\\\.)+[\\\\w-]{2,4}$/\` but received \\"\\"",
      "ref": Object {
        "name": "email",
      },
      "type": "string",
    },
    "enabled": Object {
      "message": "Expected a value of type \`boolean\`, but received: \`undefined\`",
      "ref": undefined,
      "type": "boolean",
    },
    "like": Array [
      Object {
        "id": Object {
          "message": "Expected a number, but received: \\"z\\"",
          "ref": undefined,
          "type": "number",
        },
        "name": Object {
          "message": "Expected a string, but received: undefined",
          "ref": undefined,
          "type": "string",
        },
      },
    ],
    "password": Object {
      "message": "Expected a string matching \`/^[a-zA-Z0-9]{3,30}/\` but received \\"___\\"",
      "ref": Object {
        "name": "password",
      },
      "type": "string",
    },
    "repeatPassword": Object {
      "message": "Expected a value of type \`Password\`, but received: \`undefined\`",
      "ref": undefined,
      "type": "Password",
    },
    "tags": Object {
      "message": "Expected an array value, but received: undefined",
      "ref": undefined,
      "type": "array",
    },
    "username": Object {
      "message": "Expected a string, but received: undefined",
      "ref": Object {
        "name": "username",
      },
      "type": "string",
    },
  },
  "values": Object {},
}
`;

exports[`superstructResolver should return all the errors from superstructResolver when validation fails with \`validateAllFieldCriteria\` set to true 1`] = `
Object {
  "errors": Object {
    "birthYear": Object {
      "message": "Expected a number, but received: \\"birthYear\\"",
      "ref": undefined,
      "type": "number",
      "types": Object {
        "number": "Expected a number, but received: \\"birthYear\\"",
      },
    },
    "email": Object {
      "message": "Expected a string matching \`/^[\\\\w-\\\\.]+@([\\\\w-]+\\\\.)+[\\\\w-]{2,4}$/\` but received \\"\\"",
      "ref": Object {
        "name": "email",
      },
      "type": "string",
      "types": Object {
        "string": "Expected a string matching \`/^[\\\\w-\\\\.]+@([\\\\w-]+\\\\.)+[\\\\w-]{2,4}$/\` but received \\"\\"",
      },
    },
    "enabled": Object {
      "message": "Expected a value of type \`boolean\`, but received: \`undefined\`",
      "ref": undefined,
      "type": "boolean",
      "types": Object {
        "boolean": "Expected a value of type \`boolean\`, but received: \`undefined\`",
      },
    },
    "like": Array [
      Object {
        "id": Object {
          "message": "Expected a number, but received: \\"z\\"",
          "ref": undefined,
          "type": "number",
          "types": Object {
            "number": "Expected a number, but received: \\"z\\"",
          },
        },
        "name": Object {
          "message": "Expected a string, but received: undefined",
          "ref": undefined,
          "type": "string",
          "types": Object {
            "string": "Expected a string, but received: undefined",
          },
        },
      },
    ],
    "password": Object {
      "message": "Expected a string matching \`/^[a-zA-Z0-9]{3,30}/\` but received \\"___\\"",
      "ref": Object {
        "name": "password",
      },
      "type": "string",
      "types": Object {
        "string": "Expected a string matching \`/^[a-zA-Z0-9]{3,30}/\` but received \\"___\\"",
      },
    },
    "repeatPassword": Object {
      "message": "Expected a value of type \`Password\`, but received: \`undefined\`",
      "ref": undefined,
      "type": "Password",
      "types": Object {
        "Password": "Expected a value of type \`Password\`, but received: \`undefined\`",
      },
    },
    "tags": Object {
      "message": "Expected an array value, but received: undefined",
      "ref": undefined,
      "type": "array",
      "types": Object {
        "array": "Expected an array value, but received: undefined",
      },
    },
    "username": Object {
      "message": "Expected a string, but received: undefined",
      "ref": Object {
        "name": "username",
      },
      "type": "string",
      "types": Object {
        "string": "Expected a string, but received: undefined",
      },
    },
  },
  "values": Object {},
}
`;
